DOCSP-50472: schema validation (#3400)
jenssegers/laravel-mongodb: docs/eloquent-models/schema-builder.txt

`Database: Migrations <https://laravel.com/docs/{+laravel-docs-version+}/migrations>`__
in the Laravel documentation.

.. _laravel-eloquent-schema-validation:

Implement Schema Validation
---------------------------

You can use the ``jsonSchema()`` method to implement :manual:`schema
validation </core/schema-validation/>` when using the following schema
builder methods:

- ``Schema::create()``: When creating a new collection

> - ``Schema::table()``: When updating collection properties

Reviewer: Q: can you only implement schema validation when using these methods for these specific purposes? or is the info after the colons just a reminder of the method's purpose?

Reviewer 2: These are the only two methods in Laravel's `Schema` that accept a callback to alter columns, so you can only implement schema validation on either creating a collection or updating it.

After you implement schema validation, the server allows you to run only

> those write operations which follow the validation rules. Use schema

Reviewer: [that/which](https://www.mongodb.com/docs/meta/style-guide/writing/use-that-which-correctly/)

```suggestion
those write operations that follow the validation rules. Use schema
```

validation to restrict data types and value ranges of document fields in

> a specified collection.

Reviewer: S: i wonder if the connection between these two sentences would be clearer if you switched them. something like:

```suggestion
You can use schema validation to restrict data types and value ranges of document fields in
a specified collection. After you implement schema validation, the server allows you to run only
those write operations which follow the validation rules. 
```

Author: makes sense

Before creating a collection with schema validation rules, you must
define a JSON schema.

The schema is a JSON object that contains key-value pairs
specifying the validation rules for your collection. At the top level,

> this object must include a ``$jsonSchema`` object. The ``$jsonSchema``

Reviewer 2: `$jsonSchema` doesn't need to be included in the schema argument (see the example). Since the method is `Blueprint::jsonSchema`, the `'$jsonSchema'` operation is already implied. 

So this documentation may be valid for MongoDB server but not for Laravel usage

Reviewer 2: Oh also! Since this is PHP, the schema argument is not an object, but an associative array.

> Before creating a collection with schema validation rules, you must
> define a JSON schema.
> 
> The schema is a JSON object that contains key-value pairs
> specifying the validation rules for your collection. At the top level,
> this object must include a ``$jsonSchema`` object. The ``$jsonSchema``
> object can include the following fields:

Reviewer: S: break up paragraphs differently. could also then remove first 'JSON'

```suggestion
Before creating a collection with schema validation rules, you must
define the schema.
A schema is a JSON object that contains key-value pairs
specifying the validation rules for your collection.
At the top level, this object must include a ``$jsonSchema`` object. The ``$jsonSchema``
object can include the following fields:
```

Author: Removed this paragraph per PV suggestion

> - ``title``: Sets an optional title for the schema.

Reviewer 2: Should we actually include this documentation? Seems to me the `$jsonSchema` operation is covered by the Mongo server docs. The Laravel ORM passes the schema through as is. I think specifying allowed fields here puts us in a situation of having to update this documentation when Mongo server specs change, when the user could just use the Mongo server docs as the single source of truth for jsonSchema usage.

Author: That makes sense - I'll cut out some of this text and apply the suggestions from your prev comment.

> - ``title``: Sets an optional title for the schema.
> - ``required``: Specifies a list of required fields for each document in your collection.
> - ``properties``: Sets property requirements for individual fields.

Reviewer: I: no periods

```suggestion
- ``title``: Sets an optional title for the schema
- ``required``: Specifies a list of required fields for each document in your collection
- ``properties``: Sets property requirements for individual fields
```

Author: removed this text

To view a full list of JSON schema object fields, see :manual:`JSON Schema
</reference/operator/query/jsonSchema/#json-schema>` in the {+server-docs-name+}.

You can optionally pass the following parameters to ``jsonSchema()``:

- ``validationLevel``: Sets the level of validation enforcement.
  Accepted values are ``"strict"`` (default) and ``"moderate"``.
- ``validationAction``: Specifies the action to take when invalid
  operations are attempted. Accepted values are ``"error"`` (default) and
  ``"warn"``.

This example demonstrates how to pass a JSON schema to the
``jsonSchema()`` method when creating a collection. The schema
validation specifies that documents in the ``pilots`` collection must

> contain the ``license_number`` field with an integer value between

Reviewer: S:

```suggestion
contain the ``license_number`` field, and the field must have an integer value between
```

Author: changed to a list
